File: HW1/cs333_assignment1/DepthFinder/java/GameBoard.java
import java.io.*;
import java.util.*;

public class GameBoard{

    public int size;
    protected int minLocation;
    protected double minValue;
    protected int queriesRemaining;
    protected double[] board;

    GameBoard(int boardsize){
	Random rand = new Random();
        this.size = boardsize;
        this.queriesRemaining = 3*(int) Math.ceil(Math.log(boardsize) / Math.log(2));	
        this.board = new double[boardsize];
        this.minLocation = rand.nextInt(boardsize-4)+2;
        this.minValue = rand.nextInt(boardsize*boardsize*boardsize)+10;
        this.board[0] = 1;
        this.board[1] = 2;
        this.board[boardsize-1] = 3;
        this.board[boardsize-2] = 4;
    }

    public String toString(){
        String s = "[";
        for (double d : this.board){
            if (d > 0){s += d + ",";}
            else{s += "??,";}
        }
        return s+"]";
    }

    public double ping(int location){
        --this.queriesRemaining;
        if (this.queriesRemaining < 0){
            System.out.println("out of pings");
            return 0;
        }
        if (this.board[location] > 0){return this.board[location];}
        if (location == this.minLocation){
            this.board[location] = this.minValue;
            return this.board[location];
        }
        double depth = 0;
        int pinged_left = 1;
        for (int i = 0; i < this.size; i++){
            depth = this.board[i];
            if (i < location && depth > 0){
                pinged_left = i;
            }
        }
        depth = 0;
        int pinged_right = this.size-2;
        for (int i = this.size-1; i > 0; i--){
            depth = this.board[i];
            if (i > location && depth > 0){
                pinged_right = i;
            }
        }
        if (location < this.minLocation){
	    if (pinged_right > this.minLocation){
                this.board[location] = (this.board[pinged_left] + this.minValue) / 2;
	    }
            else{
                this.board[location] = (this.board[pinged_left] + this.board[pinged_right]) / 2;
	    }

        }
        else{
	    if (pinged_left < this.minLocation){
                this.board[location] = (this.board[pinged_right] + this.minValue) / 2;
	    }
	    else{
                this.board[location] = (this.board[pinged_left] + this.board[pinged_right])/2;
	    }
	}
        return this.board[location];
    }

    public void finalAnswer(int location){
        if (this.queriesRemaining < 0){
            System.out.println("out of pings");
        }
	else if (this.board[location-1] <= 0 || this.board[location] <= 0 || this.board[location+1] <= 0) {
            System.out.println("neighbors not queried");
        }
        else if (this.board[location-1] < this.board[location] && this.board[location] > this.board[location+1]){
            System.out.println("trench found!");
        }
        else{
            System.out.println("not a trench");
        }
    }
}

File: HW1/cs333_assignment1/DepthFinder/python/gameboard.py
import random
import math


class GameBoard:
    def __init__(self, boardsize):
        self.size = boardsize
        self._queries_remaining = 3 * math.ceil(math.log(self.size, 2))
        self._board = [0 for i in range(self.size)]
        self._board[0] = 1
        self._board[1] = 2
        self._board[-1] = 3
        self._board[-2] = 4
        self._minlocation = random.randrange(2, boardsize-2)
        self._minvalue = random.randint(10, self.size**3)

    def __str__(self):
        s = "["
        for d in self._board:
            if d > 0:
                s += str(d) + ","
            else:
                s += "??,"
        return s + "]"


    def ping(self, location):
        self._queries_remaining -= 1
        if self._queries_remaining < 0:
            print("out of queries")
            return 0
        if self._board[location] > 0:
            return self._board[location]
        if location == self._minlocation:
            self._board[location] = self._minvalue
            return self._board[location]
        pinged_left = 1
        for i in range(len(self._board)):
            depth = self._board[i]
            if i < location and depth > 0:
                pinged_left = i
        pinged_right = self.size-2
        for i in range(len(self._board) - 1, -1, -1):
            depth = self._board[i]
            if i > location and depth > 0:
                pinged_right = i
        if location < self._minlocation:
            if pinged_right > self._minlocation:
                self._board[location] = (self._board[pinged_left] + self._minvalue) / 2
            else:
                self._board[location] = (self._board[pinged_left] + self._board[pinged_right]) / 2
        elif location > self._minlocation:
            if pinged_left < self._minlocation:
                self._board[location] = (self._board[pinged_right] + self._minvalue) / 2
            else:
                self._board[location] = (self._board[pinged_left] + self._board[pinged_right]) / 2
        return self._board[location]

    def final_answer(self, location):
        if (self._queries_remaining < 0):
            print("out of pings")
        elif (self._board[location - 1] <= 0) or (self._board[location] <= 0) or (
                self._board[location + 1] <= 0):
            print("neighbors not queried")
        elif (self._board[location - 1] < self._board[location] > self._board[location + 1]):
            print("trench found!")
        else:
            print("not a trench")
